wasm.md

```diff
@@ -1011,6 +1016,58 @@ The maximum of table size is 2^32 bytes.
     rule #maxTableSize()  => 4294967296
 ```
 
+Initializers
+------------
+
+The `elem` and `data` initializers take an offset, which is an instruction.
+This is not optional.
+The offset can either be specified explicitly with the `offset` key word, or be a single instruction.
+
+```k
+    syntax Offset ::= "(" "offset" Instr ")"
+                    | Instr
+```
+
+### Data Segments
+
+Memories can be initialized with data, specified as a list of bytes together with an offset.
+The `data` initializer simply puts these bytes into the specified memory, starting at the offset.
+
+```k
+    syntax Instr ::= Data
+    syntax Data ::= "(" "data"     TextFormatIdx Offset DataStrings ")"
+                  | "(" "data"                   Offset DataStrings ")"
+                  |     "data" "{" TextFormatIdx        DataStrings "}"
+ // -------------------------------------------------------------------
+    // Default to memory 0.
+    rule <k> ( data       OFFSET      STRINGS ) =>     ( data 0 OFFSET STRINGS ) ... </k>
+    rule <k> ( data MEMID IS:Instr    STRINGS ) => IS ~> data { MEMID  STRINGS } ... </k>
+    rule <k> ( data MEMID (offset IS) STRINGS ) => IS ~> data { MEMID  STRINGS } ... </k>
+
+    // For now, deal only with memory 0.
+    rule <k> data { MEMIDX STRING } => . ... </k>
+         <valstack> < i32 > OFFSET : STACK => STACK </valstack>
+         <memIndices> MEMIDX |-> ADDR </memIndices>
+         <memInst>
+           <mAddr> ADDR </mAddr>
+           <mdata>   DATA
+                  => #clearRange(DATA, OFFSET, OFFSET +Int #dataStringsLength(STRING) -Int 1) [ OFFSET := #dataStrings2int(STRING)]
+           </mdata>
+           ...
+         </memInst>
+
+    syntax Int ::= #lengthDataPages ( DataStrings ) [function]
+ // ----------------------------------------------------------
+    rule #lengthDataPages(DS:DataStrings) => #dataStringsLength(DS) /ceilInt #pageSize()
+
+    syntax Int ::= Int "/ceilInt" Int [function]
+ // --------------------------------------------
+    rule I1 /ceilInt I2 => (I1 /Int I2)
+      requires I1 modInt I2  ==Int 0
+    rule I1 /ceilInt I2 => (I1 /Int I2) +Int 1
+      requires I1 modInt I2 =/=Int 0
```

Comment:
let's do one rule/line here, line up the `requires` clauses.

Reply:
Also I'm not in love with using `modInt` here, as it's historically been not so great when passed to Z3.

In KEVM, we have the following operator:

```
    syntax Int ::= Int "up/Int" Int [function] 
 // ------------------------------------------ 
    rule I1 up/Int 0  => 0 
    rule I1 up/Int 1  => I1 
    rule I1 up/Int I2 => (I1 +Int (I2 -Int 1)) /Int I2 requires I2 >Int 1 
```

I'll leave it up to you whether to change this, since I don't see it causing problems yet and `Data` strings should be concrete for the most part (since they are part of the input program).

Reply:
Good one. I'd rather reuse the KEVM one.

Reply:
Is the second rule of any use? It should be the exact same as the third when `I2 == 1`